📖 Variable discovery proposal (#48)
killianmuldoon/cluster-api: docs/book/src/tasks/experimental-features/runtime-sdk/implement-topology-mutation-hook.md

* External patches can use external data (e.g. from cloud APIs) during patch generation.
* External patches can be easily reused across ClusterClasses.

> ## External variables

Reviewer: ```suggestion
## External variable definitions
```
nit

The DiscoverVariables hook can be used to supply variable definitions for use in external patches. These variable definitions are added to
any applicable ClusterClasses. Clusters using the ClusterClass can then set values for those variables.

### External variable discovery in the ClusterClass
External variable definitions are discovered by calling the DiscoverVariables runtime hook. This hook is called from the ClusterClass reconciler.
Once discovered the variable definitions are validated and stored in ClusterClass status.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
# metadata

> spec:

Reviewer: Should we also add a "global" inline variable to this example (and the Cluster below) to show that they can be used as is (i.e. we are downwards compatible there)

    # Inline variable definitions
    variables:
    - name: http-proxy
      schema:
        openAPIV3Schema:
          type: string
          default: "proxy.example.com"
          example: "proxy.example.com"
          description: "proxy for http calls."
    # External patch definitions.
    patches:
    - name: lbImageRepository
      external:
          generateExtension: generate-patches.k8s-upgrade-with-runtimesdk
          validateExtension: validate-topology.k8s-upgrade-with-runtimesdk
          ## Explicitly call variable discovery for this patch.
          variableDiscovery: variable-discovery.k8s-upgrade-with-runtimesdk
status:
    # observedGeneration is used to check that the current version of the ClusterClass is the same as that when the Status was previously written.
    # if metadata.generation isn't the same as observedGeneration Cluster using the ClusterClass should not reconcile.
    observedGeneration: xx
    # variables contains a list of all variable definitions, both inline and external, that belong to the ClusterClass.

>     variables: 

Reviewer: I think I would group by variable name. Should be the most useful for users. I think building the list that way shouldn't require significant more complicated code

>         - name: http-proxy

Reviewer: nit: indent

          schema:
            openAPIV3Schema:
             type: string
             # default is different so this variable does not match the inline definition.
             default: "proxy.example3.com"
             example: "proxy.example.com"
             description: "proxy for http calls."
          # namespace is added as this variable's schema isn't equivalent to the inline definition.

>           namespace: lbImageRepository

Reviewer: I think namespace would fit better. I assume whatever we choose here is the same that users would use when defining the variable value?

origin seems a bit strange there

>           namespace: lbImageRepository

Reviewer: Regarding: 1 entry per variable definition.

I think I would go with that. Otherwise the question is when do we merge entries together. Our current definition of "equivalent" wouldn't work as this means that we would have to drop description and example here.

          # A variable can have multiple origins (altenate name /sources/??)

>           # TODO: What purpose does this serve?

Reviewer: Yup. I have the same question.

I would have thought we just have the namespace field for ~ that

         # origins: 
         #  VariableDiscovery:
         #  - patch: lbImageRepository # not sure if to use this or variableDiscovery << to list
        - name: http-proxy
          schema:
          openAPIV3Schema:
            type: string
            default: "proxy.example.com"
            example: "proxy.example.com"
            description: "proxy for http calls."
          namespace: inline
```

### Variable namespacing
Variable definitions can be inline in the ClusterClass or from any number of external DiscoverVariables hooks.
If all variables that share a name have equivalent schemas the variables are considered to be in the global namespace.
Note: This is always the case if there is only one variable with a particular name.
The CAPI components will consider variable definitions to be equivalent when they share a name and `openAPIV3Schema` fields
excluding `description` and `example` are exactly equal. (TODO: Clarify this.)

> If two or more variables which share a name do not have equivalent schemas, the variables will be namespaced. This means they will

Reviewer: If they are equivalent namespace will be just empty and thus signal that it's in the global namespace aka not-namespaced, right?

Author: TBD 😄

appear namespaced in the ClusterClass `.status.variables` and setting values in the Cluster `.topology.spec.variables` will also require setting a namespace.

> The namespace on a variable will depend on its source (TODO: RUNTIME_EXTENSION name or patch name?)

Reviewer: I would go with patch name. What is the Runtime Extension name? (i.e. where would we get it from?)

Using the patch name has the advantage that they must be already unique and that the namespace name is under controller of the ClusterClass author

> Note: When namespaced inline variables can be set without a namespace as they are always in the global namespace.

Reviewer: I'm not sure if I would do this. This makes transitions from global<=>namespaced and back very magical

I think I would keep it simple and say:
* if a variable definition is namespaced variable values must be as well
* if a variable is global, namespace won't be set in the ClusterClass.status variable definitions and in the variable values in Cluster.spec.topology.variables

Reviewer 2: +1 to Stefan's suggestion. That would make it simple. 
All inline variables implicitly becoming global is confusing especially if there are clashes.

Reviewer: Small correction: 

> if a variable is global, namespace won't be set in the ClusterClass.status variable definitions and in the variable values in Cluster.spec.topology.variables

Not sure about if we should set the namespace in the ClusterClass.status.variable regardless. Probably it makes sense to always set it there to keep it simpler (and less magical)

### Setting values for variables in the Cluster
Setting variables using external variables requires attention to be paid to variable namespacing, as exposed in the ClusterClass status. 
Variable values are set in Cluster `.spec.topology.variables`.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
#metadata 
spec:
    topology:
        variables:
        - name: http_proxy
          # If not defined or empty this variable is in the global namepsace.
          # is a list 
          namespaces: 
            - lbImageRepository

>             - inline

Reviewer 3: kind of thinking if it is better to have a list or a single value
also, wondering if/how we can make this list SSA-friendly (what is the listMapKey?)

Reviewer: Do you mean to make the the variables or the namespaces list SSA-friendly?

> Only applicable when +listType=map. A list of field names whose values uniquely identify entries in the list. While there can be multiple keys, listMapKey is singular because keys need to be specified individually in the Go type. The key fields must be scalars.

If we use "namespace" instead of "namespaces" it would be a scalar. But I'm not sure if the list map keys can use a field that is optional (I assume namespace or namespaces is optional).

Once discovered the variable definitions are validated and stored in ClusterClass status.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
# metadata
spec:
    # Inline variable definitions
    variables:
    # This variable is unique and can be accessed globally.
    - name: no-proxy
      schema:
        openAPIV3Schema:
          type: string
          default: "internal.com"
          example: "internal.com"
          description: "comma-separated list of machine or domain names excluded from using the proxy."
    # This variable is also defined by an external variableDiscovery hook.
    - name: http-proxy
      schema:
        openAPIV3Schema:
          type: string
          default: "proxy.example.com"
          example: "proxy.example.com"
          description: "proxy for http calls."
    # External patch definitions.
    patches:
    - name: lbImageRepository
      external:
          generateExtension: generate-patches.k8s-upgrade-with-runtimesdk
          validateExtension: validate-topology.k8s-upgrade-with-runtimesdk
          ## Call variable discovery for this patch.

>           variableExtension: variable-discovery.k8s-upgrade-with-runtimesdk

Author: I'm fine with whatever naming - but discoverExtension isn't descriptive enough IMO. I'm fine with discoverVariablesExtension though.

status:
    # observedGeneration is used to check that the current version of the ClusterClass is the same as that when the Status was previously written.
    # if metadata.generation isn't the same as observedGeneration Cluster using the ClusterClass should not reconcile.
    observedGeneration: xx
    # variables contains a list of all variable definitions, both inline and from external patches, that belong to the ClusterClass.
    variables:
      - name: no-proxy
        definitions:

>           - namespace: inline

Author: Yeah - this is what matches the existing definition, so it makes the most sense.

    observedGeneration: xx
    # variables contains a list of all variable definitions, both inline and from external patches, that belong to the ClusterClass.
    variables:
      - name: no-proxy
        definitions:
          - namespace: inline
            required: true
            schema:
              openAPIV3Schema:
                type: string
                default: "internal.com"
                example: "internal.com"
                description: "comma-separated list of machine or domain names excluded from using the proxy."
      - name: http-proxy
        definitions:
          - namespace: inline
            schema:
              openAPIV3Schema:
                type: string
                default: "proxy.example.com"
                example: "proxy.example.com"
                description: "proxy for http calls."
          - namespace: lbImageName
            schema:
              openAPIV3Schema:
                type: string
                default: "different.example.com"
                example: "different.example.com"
                description: "proxy for http calls."
```

### Variable namespacing

> Variable definitions can be inline in the ClusterClass or from any number of external DiscoverVariables hooks. The source 
> of a variable definition is recorded in the `namespaces` field in ClusterClass `.status.variables`.
> Variables that are defined by an external DiscoverVariables hook will have the name of the patch they are associated with as their namespace.
> Variables that are defined in the ClusterClass `.spec.variables` will have the namespace `inline`.
> Note: `inline` is a reserved namespace. It can not be used as the name of an external patch to avoid conflicts.
> 
> If all variables that share a name have equivalent schemas the variables are considered `global` . `global` variables can
> be set without providing a namespace - [see below](#setting-values-for-variables-in-the-cluster). The CAPI components will
> consider variable definitions to be equivalent when they share a name and their schema is exactly equal.
> 
> ### Setting values for variables in the Cluster
> Setting variables that are defined with external variable definitions requires attention to be paid to variable namespacing, as exposed in the ClusterClass status. 
> Variable values are set in Cluster `.spec.topology.variables`.
> 
> ```yaml
> apiVersion: cluster.x-k8s.io/v1beta1
> kind: Cluster
> #metadata 
> spec:
>     topology:
>         variables:
>         - name: no-proxy
>           value: "internal.domain.com"
>           # namespace is not needed as this variable is in the global namespace.
>         - name: http-proxy
>           namespace: inline
>           value: "proxy.example2.com"
>         - name: http-proxy
>           namespace: lbImageRepository
>           value: "proxy.example2.com"
> ```

Reviewer 3: what about
```suggestion
Variable definitions can be inline in the ClusterClass or the result of calls to  DiscoverVariables hooks. 
The system provides support for handling the same variable being defined in multiple places, and this support is designed around two principles:
- Keeping user UX as simple as possible.  
   The user just wants to provide the value for each variables once, they should not be bothered by the fact that this same variable is defined more than once or used in different ways.
- ClusterClass authors and Runtime Extension authors are responsible for acting proactively in order to ensure the best UX for the users. If they fail in doing so the user UX will degrade accordingly

I practice this means:
- All variables definitions will be assigned to a namespace:
   - Variable definitions defined in the ClusterClass are assigned to the inline namespace.  
   - Variable definitions resulting from external DiscoverVariable hooks are associated with a namespace having the same name as the patch defining the hook call
   - NOTE: `.status.variables` field in ClusterClass contains the result of this process
   - NOTE: `inline` becomes a reserved word that cannot be used anymore as a patch name.
   
 - If all variables definitions that share a name have equivalent schemas the definitions are considered `global`
    - Users can provide a single value (without namespace) and this will be used for all the corresponding definitions
    - NOTE: Variable definitions are equivalent when they share a name and their schema is exactly equal.  
 - If variables that share a name do not have equivalent schemas they are not `global`. 
    - Users must provide a single value for each variable definition, identified by the namespace (degraded UX).
    

Below is an example of how to set global and namespaced variables in a Cluster:

Setting variables using external variables requires attention to be paid to variable namespacing, as exposed in the ClusterClass status. 
Variable values are set in Cluster `.spec.topology.variables`.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
#metadata 
spec:
    topology:
        variables:
        # namespace is not needed as this variable is global; this also allows migration of existing variables.
        - name: no-proxy
          value: "internal.domain.com"
        # namespaced variables (the value must be provided twice according the two different schemas.
        - name: http-proxy
          namespaces: patch1
          value: http://proxy.example2.com:1234
       - name: http-proxy
          namespaces: patch2
          value:
             host: proxy.example2.com
             port: 1234
```

```

so we document the why we are doing all this work; also changed the example a little bit, PTAL

Author: IMO there's too much detail here around the reasoning for this - frontrunning potential comments. I think it's better to just state what users should do, and what situations it covers instead of principles in this doc - given it's for implementers.

If you think this is really useful though maybe it would fit better in the proposal?

Reviewer 3: ACK, and I don't want to block it for the first iteration
just please take a look at the example, if it makes sense to use it (it makes it more explicit that schemas are not equal)

Author: I did take the example :smile: - it really does make it more clear.

* External patches can use external data (e.g. from cloud APIs) during patch generation.
* External patches can be easily reused across ClusterClasses.

## External variable definitions
The DiscoverVariables hook can be used to supply variable definitions for use in external patches. These variable definitions are added to
the status of any applicable ClusterClasses. Clusters using the ClusterClass can then set values for those variables.

### External variable discovery in the ClusterClass
External variable definitions are discovered by calling the DiscoverVariables runtime hook. This hook is called from the ClusterClass reconciler.
Once discovered the variable definitions are validated and stored in ClusterClass status.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
# metadata
spec:
    # Inline variable definitions
    variables:
    # This variable is unique and can be accessed globally.

>     - name: no-proxy

Reviewer: ```suggestion
    - name: no-proxy
      required: true
```
for consistency with status

    variables:
      - name: no-proxy
        definitions:
          - namespace: inline
            required: true
            schema:
              openAPIV3Schema:
                type: string
                default: "internal.com"
                example: "internal.com"
                description: "comma-separated list of machine or domain names excluded from using the proxy."
      - name: http-proxy
        definitions:
          - namespace: inline
            schema:
              openAPIV3Schema:
                type: string
                default: "proxy.example.com"
                example: "proxy.example.com"
                description: "proxy for http calls."
          - namespace: lbImageName
            schema:
              openAPIV3Schema:
                type: string
                default: "different.example.com"
                example: "different.example.com"
                description: "proxy for http calls."
```

### Variable namespacing
Variable definitions can be inline in the ClusterClass or from any number of external DiscoverVariables hooks. The source 

> of a variable definition is recorded in the `namespaces` field in ClusterClass `.status.variables`.

Reviewer: ```suggestion
of a variable definition is recorded in the `namespace` field in ClusterClass `.status.variables`.
```

Variables that are defined in the ClusterClass `.spec.variables` will have the namespace `inline`.
Note: `inline` is a reserved namespace. It can not be used as the name of an external patch to avoid conflicts.

If all variables that share a name have equivalent schemas the variables are considered `global` . `global` variables can
be set without providing a namespace - [see below](#setting-values-for-variables-in-the-cluster). The CAPI components will
consider variable definitions to be equivalent when they share a name and their schema is exactly equal.

### Setting values for variables in the Cluster
Setting variables that are defined with external variable definitions requires attention to be paid to variable namespacing, as exposed in the ClusterClass status. 
Variable values are set in Cluster `.spec.topology.variables`.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
#metadata 
spec:
    topology:
      variables:
        # namespace is not needed as this variable is global.
        - name: no-proxy
          value: "internal.domain.com"
        # namespaced variables require values for each individual schema.
        - name: http-proxy

>           namespaces: patch1

Reviewer: ```suggestion
          namespace: patch1
```

          value: http://proxy.example2.com:1234
        - name: http-proxy

>           namespaces: patch2

Reviewer: ```suggestion
          namespace: patch2
```

  always return the same error message. Otherwise the system might became unstable due to controllers being overloaded
  by continuous changes to Kubernetes resources as these messages are reported as conditions. See [error messages](implement-extensions.md#error-messages).

### Variable discovery guidelines

> * **Unique variable names**: When using external variable discovery, care should be taken to ensure that variable names are unique.

Reviewer: This sentence refers to variables returned by a single call or across external patches & inline variables?

Author: Replaced with Fabrizio's version, but this meant across all patches initially.

  If the names are not unique the ClusterClass controller will attempt to deduplicate variables by comparing their definition. If this
  is not possible the variables will be namespaced and must be referenced by their namespace when defined in a Cluster.

> * **Avoid updating variable definitions**: When using external variables updating the definitions of variables in use by existing Clusters
>   can cause reconciliation to fail where existing variable values are no longer valid, or where changing the definitions results in namespacing of
>   variables which were previously not namespaced.

Reviewer 3: What about
```suggestion
* **Avoid breaking changes variable definitions**: Changing a variable definition can lead to problems on existing clusters because reconciliation will stop if variable values will not match anymore the new definition.
Furthermore, when the variables defined for this patch are going to be used in combinations with other patches providing variables with the same name, changes to variable definition can lead to the system requiring the users to provide different values for each namespace the value exists in.
As a consequence updates to the variable definition should always be carefully evaluated, and very well documented in the extension release notes, so ClusterClass authors can evaluate impacts of changes before upgrade.
```

}
</script>

### DiscoverVariables

> A DiscoverVariables call returns definitions for one or more variables. This variable definition is the same as that used for inline variables.

Reviewer 3: ```suggestion
A DiscoverVariables call returns definitions for one or more variables. 
```

> Example variable definition:

Reviewer 3: Let's drop the example, we already have it down in the following paragraphs

```yaml
  - name: etcdImageTag # name of the variable
    required: true # whether the variable must be defined for Clusters.
    schema:
      openAPIV3Schema: 
        type: string 
        default: "3.5.3-0" # variable value which is defaulted during reconciliation if not defined elsewhere.
        example: "3.5.3-0" # example for user guidance.
        description: "etcdImageTag sets the tag for the etcd image." #description of variable use.
        # Additional fields from clusterv1.JSONSchemaProps
```

#### Example Request:

* The request is a simple request which contains the name and namespace of the ClusterClass it is called for.

> TODO: Should this send the full ClusterClass?

Reviewer 3: I think not, the name should be ok.
Also, I think we are missing the settings

Reviewer 2: Echoing what Fabrizio said, we should add settings here. Settings is part of all runtime extension hook request objects.

Reviewer: As mentioned here: https://github.com/killianmuldoon/cluster-api/pull/48/files#r1065849823

I"m not sure if we should send the ClusterClass name. I think we facilitate some strange use cases this way.

But no strong objection from my side if we think it's a good idea.

> TODO: Should this send the full ClusterClass?

Reviewer: I would start simple. I probably wouldn't even send the ClusterClass. I'm not sure if we should facilitate that folks are coupling their external patch extension to closely to a specific ClusterClass (either by name or by get'ing the ClusterClass - which can always be out of sync).

If we don't have a use case for folks accesing the ClusterClass from the external patch extension I wouldn't add any fields for now.

They can still do some parameterization via settings.

```yaml
apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
kind: DiscoverVariablesRequest
clusterClass:
- name: <clusterClass-name>
  namespace: <clusterClass-namespace>
```

#### Example Response:

```yaml
apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
kind: DiscoverVariablesResponse

> status: Success # or Failure

Reviewer 2: The response should also include `message`. `status` and `message` are required for all response objects.

Reviewer: Hm definitely of the API type. But are we actually setting message if the status is Success?

I think originally we planned to only set the message if we have a Failure

Reviewer: Hm our lifecycle handlers don't even set the status in the success case. That doesn't seem right :)

}
</script>

### DiscoverVariables

A DiscoverVariables call returns definitions for one or more variables.

#### Example Request:

* The request is a simple request which contains the name and namespace of the ClusterClass it is called for.
TODO: Should this send the full ClusterClass?

```yaml
apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
kind: DiscoverVariablesRequest

> settings: <Runtime Extension settings>

Author: Will do

}
</script>

### DiscoverVariables

A DiscoverVariables call returns definitions for one or more variables.

#### Example Request:

> * The request is a simple request which contains the name and namespace of the ClusterClass it is called for.

Reviewer: It doesn't contain the name/namespace anymore
